lxd/cgroup/abstraction.go

```diff
@@ -1027,7 +1030,14 @@ func (cg *CGroup) GetIOStats() (map[string]*IOStats, error) {
 		scanner := bufio.NewScanner(strings.NewReader(val))
 
 		for scanner.Scan() {
-			devID, stats, _ := strings.Cut(scanner.Text(), " ")
+			var devID, nestedDevIDs string
+			// There might be multiple devID in the beginning of the line, we need to take the rightmost one.
```

Comment:
I think we can do this without using regex and just make a small change by parsing the line as space separated values earlier:

```go
var devID string
var ioStats IOStats{}

			for _, statPart := range strings.Split(scanner.Text(), " ") {
               if strings.Contains(statPart, ":") {
                  // Store the last dev ID as this works around a kernel bug where multiple dev IDs could appear on a single line.
                  devID = statPart
                  continue
               }

                // Parse the stat value.
				statName, statValueStr, found := strings.Cut(statPart, "=")
				if !found {
					return nil, fmt.Errorf("Failed extracting io.stat %q (from %q)", statPart, scanner.Text())
				}

				statValue, err := strconv.ParseUint(statValueStr, 10, 64)
				if err != nil {
					return nil, fmt.Errorf("Failed parsing io.stat %q %q (from %q): %w", statName, statValueStr, scanner.Text(), err)
				}

				switch statName {
				case "rbytes":
					ioStats.ReadBytes = statValue
				case "wbytes":
					ioStats.WrittenBytes = statValue
				case "rios":
					ioStats.ReadsCompleted = statValue
				case "wios":
					ioStats.WritesCompleted = statValue
				}
			}

			ioMap[partMap[devID]] = ioStats
		}
```

Reply:
CC @simondeziel @monstermunchkin

Reply:
Looks good to me.

Reply:
LGTM

Reply:
I've updated my PoC code to handle lines with device IDs but no metrics.